src/toxcore/dht/server/ping.rs

```diff
@@ -0,0 +1,236 @@
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Module for PingRequest
+*/
+
+use std::time::{Duration, Instant};
+use std::mem;
+
+use futures::{future, stream, Stream};
+use parking_lot::RwLock;
+
+use toxcore::dht::packed_node::*;
+use toxcore::dht::kbucket::*;
+use toxcore::dht::server::*;
+use toxcore::dht::dht_node::*;
+use toxcore::io_tokio::IoFuture;
+
+/// Hold data for sending PingRequest
+pub struct Ping {
+    last_time_send_ping: Instant,
+    nodes_to_send_ping: RwLock<Bucket>,
+}
+
+impl Ping {
+    /// new Ping object
+    pub fn new() -> Self {
+        Ping {
+            last_time_send_ping: Instant::now(),
+            nodes_to_send_ping: RwLock::new(Bucket::new(None)),
+        }
+    }
+
+    fn is_bad_node_timed_out(node: &DhtNode, server: &Server) -> bool {
+        node.last_resp_time.elapsed() > Duration::from_secs(server.config.bad_node_timeout)
+    }
+
+    fn is_friend(node: &PackedNode, server: &Server) -> bool {
+        server.friends.read().iter().any(|friend| friend.pk == node.pk)
+    }
+
+    fn is_in_close_list(node: &PackedNode, server: &Server) -> bool {
+        server.friends.read().iter()
+            .map(|friend| friend.close_nodes.nodes.iter().any(|peer| peer.pk == node.pk))
+            .any(|result_bool| result_bool)
+    }
+
+    fn is_in_ping_list(node: &PackedNode, ping_list: &Bucket) -> bool {
+        ping_list.nodes.iter().any(|peer| peer.pk == node.pk)
+    }
+
+    fn is_iterate_time(&self, iterate_interval: Duration) -> bool {
+        self.last_time_send_ping.elapsed() >= iterate_interval
+    }
+
+    /// try to add node to list to send PingRequest
+    /// return true if node is added, false otherwise
+    pub fn try_add(&mut self, server: &Server, node: &PackedNode) -> bool {
+        // if node already exists in close list and not timed out, then don't send PingRequest
+        match server.close_nodes.read().find_node(&node.pk) {
+            Some(ref node_in_close_list) if !Ping::is_bad_node_timed_out(node_in_close_list, server) => return false,
+            _ => {},
+        };
+
+        // if node is not addable to close list, dont't send PingRequest
+        if !server.close_nodes.read().can_add(node) {
+            return false
+        }
+
+        // If node is friend and don't exist in friend's close list then send PingRequest
+        if Ping::is_friend(node, server) && !Ping::is_in_close_list(node, server) {
+            server.send_ping_req(node);
+            return false
+        }
+
+        let mut nodes_to_send_ping = self.nodes_to_send_ping.write();
+        // if node already exists in ping list, then don't add
+        if Ping::is_in_ping_list(node, nodes_to_send_ping.deref()) {
+            return false
+        }
+
+        // PingRequest is sent only for maximum 8 nodes in Bucket
+        nodes_to_send_ping.try_add(&server.pk, node)
+    }
+
+    /// send PingRequest to all nodes in list
```

Comment:
Let it be clear: do we send only 8 pings? And we do it only once since you used mem::swap. Am I correct?

Reply:
Yes.
ping list is Bucket. So only 8 closer nodes will receive PingReqeust.
Adding node to Bucket and sending Ping is separate jobs.
We check ping Bucket to send Ping every 2 second.